### PR TITLE
Add implementation for Intel oneAPI

### PR DESCRIPTION
**Pull request by fwyzard**
This is not optimised, but at least it runs.

Present status:
  - using the OpenCL CPU backend it gives the correct results.
  - using the OpenCL "neo" GPU backend it needs a couple of workarounds to avoid aborting or segfaulting, and still gives an incorrect result.

### CODE REVIEW

File: rawtodigi_oneapi.cc [hunk 1]
```diff
@@ -0,0 +1,507 @@
...
+
+    {
+      queue.submit(
+        [&](cl::sycl::handler &cgh) {
+          cl::sycl::stream out(64 * 1024, 80, cgh);
+          auto dpct_global_range = cl::sycl::range<3>(blocks, 1, 1) * cl::sycl::range<3>(threadsPerBlock, 1, 1);
+          auto dpct_local_range = cl::sycl::range<3>(threadsPerBlock, 1, 1);
+          cgh.parallel_for<dpct_kernel_name<class rawtodigi_kernel_5b0d6a>>(
+            cl::sycl::nd_range<3>(cl::sycl::range<3>(dpct_global_range.get(2), dpct_global_range.get(1), dpct_global_range.get(0)), cl::sycl::range<3>(dpct_local_range.get(2), dpct_local_range.get(1), dpct_local_range.get(0))),
+            [=](cl::sycl::nd_item<3> item_ct1) {
```
Comment (makortel) [inline rawtodigi_oneapi.cc:498]: Beautiful, already without using the buffers+handles approach.

**Review by makortel — COMMENTED** [review]
This was interesting, thanks! The overall structure indeed feels much the same as CUDA (details are of course different, especially the kernel launch).

As I mentioned in #11, I'd prefer to merge that one first, and then go into these new variants.

**Comment by fwyzard** [comment]
> This was interesting, thanks! The overall structure indeed feels much the same as CUDA (details are of course different, especially the kernel launch).

That's partly because this was initially "translated" from CUDA by the oneAPI portability tool.
It should probably be cleaned up a bit more, but I'm more worried that it segfaults with the latest Intel 
GPU driver :-(

> As I mentioned in #11, I'd prefer to merge that one first, and then go into these new variants.

Agreed.

**Comment by makortel** [comment]
#11 is now merged, could you please rebase (mostly to make the GitHub diff cleaner)?

**Comment by fwyzard** [comment]
I've reset the target branch, the diff should look good now.

**Comment by makortel** [comment]
Looks good. Could you also update the README (at least minimally if not full instructions)?

File: Makefile [hunk 2]
```diff
@@ -160,3 +162,25 @@ kokkos:
...
+ifdef ONEAPI_CXX
+oneapi: main-oneapi
+	@echo -e $(GREEN)oneAPI targets built$(RESET)
+
+oneapi-debug: debug-oneapi
+	@echo -e $(GREEN)oneAPI debug targets built$(RESET)
+
+# oneAPI implementation
+main-oneapi: main_oneapi.cc rawtodigi_oneapi.cc rawtodigi_oneapi.h
+	$(ONEAPI_CXX) -O2 -std=c++14 -DDIGI_ONEAPI -o $@ main_oneapi.cc rawtodigi_oneapi.cc
```
Comment (makortel) [inline Makefile]: Is `dpcpp` also restricted to C++14? (i.e. no C++17)

**Comment by fwyzard** [comment]
According to the Intel engineer we are in touch with, support for c++14 and
later is "in progress" during the beta phase; eventually I expect c++17 and
later to be supported, but I have not tried it yet.

**Comment by makortel** [comment]
@fwyzard Is this one ready now? (the README can be dealt with later as well)

**Comment by fwyzard** [comment]
As much as it can be, with oneAPI beta 3.
While beta 4 is expected around the end of January...

**Comment by makortel** [comment]
Let's take it in then.

**Pull request merged**